src/kernel/sporadic.c

```diff
@@ -283,18 +284,12 @@ void refill_budget_check(ticks_t usage)
     assert(sc->scPeriod > 0);
     REFILL_SANITY_START(sc);
 
-    while (unlikely(REFILL_HEAD(sc).rAmount <= usage)) {
-        /* exhaust and schedule replenishment */
-        usage -= REFILL_HEAD(sc).rAmount;
-        if (refill_single(sc)) {
-            /* update in place */
-            REFILL_HEAD(sc).rTime += sc->scPeriod;
-        } else {
-            refill_t old_head = refill_pop_head(sc);
-            old_head.rTime = old_head.rTime + sc->scPeriod;
-            schedule_used(sc, old_head);
-        }
-    }
+    /* This implementation is simply the most simple operation that
```

Comment:
whoa. What are the implications of this change? Are we going to fix it later?

You could likely rewrite the loop so it more obviously terminates, assuming that's the problem.

Reply:
The problem isn't that it terminates, termination is fairly straightforward. The issue is that charging overrun by charging time in advance can push refills into the future by `(overrun/budget)*period` and it's been difficult to determine under what circumstances that would lead to integer overflow. We need to identify an upper bound for the overrun amount (we impose the upper bound on `period` from your PR and the lower bound on `budget` results from the WCET).

Reply:
I'm struggling to understand the impact of this change on scheduling. I'm really not comfortable with it at all without taking the time to sit down and go through it line by line. Can someone explain in detail what changes? I need to know the exact implications.

Reply:
I suspect a better place to start would be on the invariants and the model used in verification. From all the discussion that has occurred there seems to have been a misunderstanding about specifically what properties we want to guarantee and many of the changes were put in place to ensure properties that it seems we don't actually want.